File: app/src/main/java/com/example/david/lists/widget/di/WidgetConfigModule.java
package com.example.david.lists.widget.di;

import android.app.Application;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.ViewModelProviders;

import com.example.david.lists.data.repository.IRepository;
import com.example.david.lists.di.view.common.ViewScope;
import com.example.david.lists.widget.configactivity.IWidgetConfigAdapter;
import com.example.david.lists.widget.configactivity.IWidgetConfigViewModel;
import com.example.david.lists.widget.configactivity.WidgetConfigAdapterImpl;
import com.example.david.lists.widget.configactivity.WidgetConfigViewModelFactory;
import com.example.david.lists.widget.configactivity.WidgetConfigViewModelImpl;

import dagger.Module;
import dagger.Provides;

@Module
final class WidgetConfigModule {
    @ViewScope
    @Provides
    IWidgetConfigViewModel viewModel(AppCompatActivity activity, ViewModelProvider.Factory factory) {
        return ViewModelProviders.of(activity, factory).get(WidgetConfigViewModelImpl.class);
    }

    @ViewScope
    @Provides
    ViewModelProvider.Factory factory(Application application, IRepository repository, int widgetId) {
        return new WidgetConfigViewModelFactory(application, repository, widgetId);
    }

    @ViewScope
    @Provides
    IWidgetConfigAdapter adapter(IWidgetConfigViewModel viewModel) {
        return new WidgetConfigAdapterImpl(viewModel);
    }
}

File: app/src/main/java/com/precopia/david/lists/widget/configview/buildlogic/WidgetConfigModule.kt
package com.precopia.david.lists.widget.configview.buildlogic

import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import com.precopia.david.lists.common.buildlogic.ViewScope
import com.precopia.david.lists.util.ISchedulerProviderContract
import com.precopia.david.lists.widget.common.UtilWidgetKeys
import com.precopia.david.lists.widget.configview.IWidgetConfigContract
import com.precopia.david.lists.widget.configview.WidgetConfigAdapter
import com.precopia.david.lists.widget.configview.WidgetConfigLogic
import com.precopia.david.lists.widget.configview.WidgetConfigViewModel
import com.precopia.domain.repository.IRepositoryContract
import dagger.Module
import dagger.Provides
import io.reactivex.rxjava3.disposables.CompositeDisposable

@Module
class WidgetConfigModule {
    @ViewScope
    @Provides
    fun logic(view: AppCompatActivity,
              factory: ViewModelProvider.NewInstanceFactory
    ): IWidgetConfigContract.Logic {
        return ViewModelProvider(view, factory).get(WidgetConfigLogic::class.java)
    }

    @ViewScope
    @Provides
    fun factory(viewModel: IWidgetConfigContract.ViewModel,
                repo: IRepositoryContract.Repository,
                schedulerProvider: ISchedulerProviderContract,
                disposable: CompositeDisposable
    ): ViewModelProvider.NewInstanceFactory {
        return WidgetConfigLogicFactory(viewModel, repo, schedulerProvider, disposable)
    }

    @JvmSuppressWildcards
    @ViewScope
    @Provides
    fun viewModel(utilWidgetKeys: UtilWidgetKeys,
                  getStringRes: (Int) -> String): IWidgetConfigContract.ViewModel {
        return WidgetConfigViewModel(utilWidgetKeys, getStringRes)
    }

    @ViewScope
    @Provides
    fun adapter(logic: IWidgetConfigContract.Logic): IWidgetConfigContract.Adapter {
        return WidgetConfigAdapter(logic)
    }
}
